views/channel-archiving.js

```diff
@@ -0,0 +1,121 @@
+const EventEmitter = require('events').EventEmitter
+const pump = require('pump')
+const Writable = require('readable-stream').Writable
+const makeView = require('kappa-view')
+  
+/*
+view data structure, the value (1) doesn't matter
+archive!<channelname>: '<pubkey of archiver>:<reason string>'
+archive!default: 'feed..b01:unused channel'
+
+ARCHIVE msg: leveldb PUT
+UNARCHIVE msg: leveldb DEL
+
+message schema: 
+{ 
+  channel: <channel>,
+  type: 'channel/archive' or 'channel/unarchive',
+  key: <pubkey of archiver>,
+  reason: <optional string reason for archiving>
+}
+*/
+
+function getAuthorizedKeys (kcore, cb) {
+  return Promise.all([getKeys('admin'), getKeys('mod')]).then(res => cb(res[0].concat(res[1]).map(row => row.id)))
+
+  // collect pubkeys for cabal-wide mods or admins
+  function getKeys (flag) {
+    return new Promise((resolve, reject) => {
+      function processResult (err, result) {
+        if (err) resolve([])
+        resolve(result)
+      }
+      kcore.api.moderation.listByFlag({ flag, channel: '@' }, processResult)
+    })
+  }
+}
+
+module.exports = function (lvl) {
+  var events = new EventEmitter()
+
+  return makeView(lvl, function (db) {
+    return {
+      maxBatch: 500,
+      map: function (msgs, next) {
+        // 1. go over each msg
+        // 2. check if it's an archive/unarchive msg (skip if not)
+        // 3. accumulate a levelup PUT or DEL operation for it
+        // 4. write it all to leveldb as a batch
+        const ops = []
+        const seen = {}
+        msgs.forEach(function (msg) {
+          if (!sanitize(msg)) { return }
+          const channel = msg.value.content.channel
+          const reason = msg.value.content.reason || ''
+          const key = msg.key
+          const pair = channel+key // a composite key, to know if we have seen this pair
+          if (msg.value.type === 'channel/archive') {
+            ops.push({
+              type: 'put',
+              key: `archive!${channel}!${key}`,
+              value: `${reason}`
+            })
+
+            if (!seen[pair]) events.emit('archive', channel, reason, key)
+            seen[pair] = true
```

Comment:
What is the logic around `seen` for?

Reply:
it wasn't really that well thoughtout, removed it–thanks!